### ISSUE TITLE
debug CLion

### ISSUE THREAD

**Issue by zzu402** [Genymobile/scrcpy#4395]
First, I need debug client with CLion in WIN10.
Then, CLion just can debug cmake project.
So, I need write a CMakeList.txt. But I'm not skilled in cmake. I need help.
This is my CMakeLists.txt, in app/src

cmake_minimum_required(VERSION 3.17)
project(scrcpy)

set(CMAKE_C_FLAGS "-std=c11")

SET(ROUT_DIR ${PROJECT_SOURCE_DIR})
SET(ANDROID_DIR ${PROJECT_SOURCE_DIR}/android)
SET(UTIL_DIR ${PROJECT_SOURCE_DIR}/util)

include_directories(${ROUT_DIR} ${ANDROID_DIR} ${UTIL_DIR})
AUX_SOURCE_DIRECTORY(${PROJECT_SOURCE_DIR} FILES)

find_path(AVFORMAT_INCLUDE_DIR libavformat/avformat.h)
find_library(AVFORMAT_LIBRARY avformat)
find_path(AVCODEC_INCLUDE_DIR libavcodec/avcodec.h)
find_library(AVCODEC_LIBRARY avcodec)
find_path(AVUTIL_INCLUDE_DIR libavutil/avutil.h)
find_library(AVUTIL_LIBRARY avutil)

find_package(sdl2 REQUIRED)

SET(THIRD_INCLUDE_DIR ${AVFORMAT_INCLUDE_DIR} ${AVCODEC_INCLUDE_DIR} ${AVUTIL_INCLUDE_DIR} ${SDL2_INCLUDE_DIRS})
INCLUDE_DIRECTORIES(${THIRD_INCLUDE_DIR})

SET(THIRD_LIBS ${AVFORMAT_LIBRARY} ${AVCODEC_LIBRARY} ${AVUTIL_LIBRARY} ${SDL2_LIBRARIES})

ADD_SUBDIRECTORY(util)

IF (CMAKE_SYSTEM_NAME MATCHES "Windows")
    set(SOURCE ${PROJECT_SOURCE_DIR}/sys/win)
    aux_source_directory(${SOURCE} FILE)

    SET(SRC ${FILES} ${FILE})
    ADD_EXECUTABLE(${PROJECT_NAME} ${SRC})
else()
    set(SOURCE ${PROJECT_SOURCE_DIR}/sys/unix)
    aux_source_directory(${SOURCE} FILE)

    SET(SRC ${FILES} ${FILE})
    ADD_EXECUTABLE(${PROJECT_NAME} ${SRC})
endif (CMAKE_SYSTEM_NAME MATCHES "Windows")

TARGET_LINK_LIBRARIES(${PROJECT_NAME} ${THIRD_LIBS} util)

# FOR CONFIG.H
SET(DEFAULT_BIT_RATE 8000000)
SET(DEFAULT_LOCAL_PORT_RANGE_FIRST 27183)
SET(DEFAULT_LOCAL_PORT_RANGE_LAST 27199)
SET(DEFAULT_MAX_SIZE 0)
SET(DEFAULT_LOCK_VIDEO_ORIENTATION -1)
SET(PREFIX \"\")
SET(SCRCPY_VERSION \"V1.16\")

OPTION(HIDPI_SUPPORT "hidpi support" ON)
OPTION(NDEBUG "build type" OFF)
OPTION(SERVER_DEBUGGER "server debugger" ON)
OPTION(SERVER_DEBUGGER_METHOD_NEW "server_debugger_method" ON)
OPTION(WINDOWS_NOCONSOLE, "WINDOWS NOCONSOLE" OFF)

configure_file (
        "${PROJECT_SOURCE_DIR}/config.h.in"
        "${PROJECT_SOURCE_DIR}/config.h"
)

This is sub CMakeLists.txt in util

CMAKE_MINIMUM_REQUIRED(VERSION 3.17)
PROJECT(util)

INCLUDE_DIRECTORIES(${PROJECT_SOURCE_DIR} ${ROOT_DIR})

AUX_SOURCE_DIRECTORY(${PROJECT_SOURCE_DIR} FILES)

ADD_LIBRARY(${PROJECT_NAME} ${FILES})

and make success，this is console info

"F:\Program\CLion 2023.2.2\bin\cmake\win\x64\bin\cmake.exe" --debug-server-port 51471 --debug-token-path C:\Users\41535\AppData\Local\Temp\cmake-debugtoken -S E:\scrcpy\scrcpy\app\src -B E:\scrcpy\scrcpy\app\src\cmake-build-debug -G Ninja -DCMAKE_BUILD_TYPE=Debug "-DCMAKE_MAKE_PROGRAM=F:/Program/CLion 2023.2.2/bin/ninja/win/x64/ninja.exe"
-- Configuring done (0.1s)
-- Generating done (0.0s)
-- Build files have been written to: E:/scrcpy/scrcpy/app/src/cmake-build-debug
Process terminated with exit code 0

But when I execute main,some error accur.

====================[ Build | scrcpy | Debug ]==================================
"F:\Program\CLion 2023.2.2\bin\cmake\win\x64\bin\cmake.exe" --build E:\scrcpy\scrcpy\app\src\cmake-build-debug --target scrcpy -j 10
[1/3] Building C object util/CMakeFiles/util.dir/thread.c.obj
FAILED: util/CMakeFiles/util.dir/thread.c.obj 
E:\msys64\mingw64\bin\cc.exe  -IE:/scrcpy/scrcpy/app/src -IE:/scrcpy/scrcpy/app/src/android -IE:/scrcpy/scrcpy/app/src/util -IE:/msys64/mingw64/include/SDL2 -std=c11 -g -fdiagnostics-color=always -MD -MT util/CMakeFiles/util.dir/thread.c.obj -MF util\CMakeFiles\util.dir\thread.c.obj.d -o util/CMakeFiles/util.dir/thread.c.obj -c E:/scrcpy/scrcpy/app/src/util/thread.c
In file included from E:/scrcpy/scrcpy/app/src/util/thread.c:5:
E:/scrcpy/scrcpy/app/src/util/thread.c: In function 'sc_thread_create':
E:/scrcpy/scrcpy/app/src/util/thread.c:16:30: error: '_beginthreadex' undeclared (first use in this function); did you mean 'SDL_beginthread'?
   16 |     SDL_Thread *sdl_thread = SDL_CreateThread(fn, name, userdata);
      |                              ^~~~~~~~~~~~~~~~
E:/scrcpy/scrcpy/app/src/util/thread.c:16:30: note: each undeclared identifier is reported only once for each function it appears in
E:/scrcpy/scrcpy/app/src/util/thread.c:16:30: error: '_endthreadex' undeclared (first use in this function)
   16 |     SDL_Thread *sdl_thread = SDL_CreateThread(fn, name, userdata);
      |                              ^~~~~~~~~~~~~~~~
ninja: build stopped: subcommand failed.

what can i do？

**Comment by rom1v** [comment]
Maybe see #2008 #2032.

**Comment by zzu402** [comment]
> Maybe see #2008 #2032.

![image](https://github.com/Genymobile/scrcpy/assets/18510098/196c623b-3819-48cd-b2a7-056f5b85a1c4)

它看起来是跟util/prcocess.h 重名了。
It looks like it has the same name as util/prcoess. h

Now I have renamed util/process. h, and another issue has arisen.，this is error info：


====================[ Build | scrcpy | Debug ]==================================
"F:\Program\CLion 2023.2.2\bin\cmake\win\x64\bin\cmake.exe" --build E:\scrcpy\scrcpy\app\src\cmake-build-debug --target scrcpy -j 10
[1/1] Linking C executable scrcpy.exe
FAILED: scrcpy.exe 
cmd.exe /C "cd . && E:\msys64\mingw64\bin\cc.exe -std=c11 -g  CMakeFiles/scrcpy.dir/audio_player.c.obj CMakeFiles/scrcpy.dir/cli.c.obj CMakeFiles/scrcpy.dir/clock.c.obj CMakeFiles/scrcpy.dir/compat.c.obj CMakeFiles/scrcpy.dir/control_msg.c.obj CMakeFiles/scrcpy.dir/controller.c.obj CMakeFiles/scrcpy.dir/decoder.c.obj CMakeFiles/scrcpy.dir/delay_buffer.c.obj CMakeFiles/scrcpy.dir/demuxer.c.obj CMakeFiles/scrcpy.dir/device_msg.c.obj CMakeFiles/scrcpy.dir/display.c.obj CMakeFiles/scrcpy.dir/file_pusher.c.obj CMakeFiles/scrcpy.dir/fps_counter.c.obj CMakeFiles/scrcpy.dir/frame_buffer.c.obj CMakeFiles/scrcpy.dir/icon.c.obj CMakeFiles/scrcpy.dir/input_manager.c.obj CMakeFiles/scrcpy.dir/keyboard_inject.c.obj CMakeFiles/scrcpy.dir/main.c.obj CMakeFiles/scrcpy.dir/mouse_inject.c.obj CMakeFiles/scrcpy.dir/opengl.c.obj CMakeFiles/scrcpy.dir/options.c.obj CMakeFiles/scrcpy.dir/packet_merger.c.obj CMakeFiles/scrcpy.dir/receiver.c.obj CMakeFiles/scrcpy.dir/recorder.c.obj CMakeFiles/scrcpy.dir/scrcpy.c.obj CMakeFiles/scrcpy.dir/screen.c.obj CMakeFiles/scrcpy.dir/server.c.obj CMakeFiles/scrcpy.dir/v4l2_sink.c.obj CMakeFiles/scrcpy.dir/version.c.obj CMakeFiles/scrcpy.dir/sys/win/file.c.obj CMakeFiles/scrcpy.dir/sys/win/process.c.obj -o scrcpy.exe -Wl,--out-implib,libscrcpy.dll.a -Wl,--major-image-version,0,--minor-image-version,0  E:/msys64/mingw64/lib/libavformat.dll.a  E:/msys64/mingw64/lib/libavcodec.dll.a  E:/msys64/mingw64/lib/libavutil.dll.a  E:/msys64/mingw64/lib/libSDL2main.a  E:/msys64/mingw64/lib/libSDL2.dll.a  util/libutil.a  -lshell32  -Wl,--undefined=WinMain  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 && cd ."
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/audio_player.c.obj: in function `sc_audio_player_frame_sink_push':
E:/scrcpy/scrcpy/app/src/audio_player.c:146:(.text+0x61c): undefined reference to `swr_get_delay'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/audio_player.c:156:(.text+0x680): undefined reference to `swr_convert'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/audio_player.c:307:(.text+0xc3b): undefined reference to `swr_set_compensation'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/audio_player.c.obj: in function `sc_audio_player_frame_sink_open':
E:/scrcpy/scrcpy/app/src/audio_player.c:367:(.text+0xeed): undefined reference to `swr_alloc'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/audio_player.c:390:(.text+0x1027): undefined reference to `swr_init'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/audio_player.c:443:(.text+0x11f2): undefined reference to `swr_free'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/audio_player.c.obj: in function `sc_audio_player_frame_sink_close':
E:/scrcpy/scrcpy/app/src/audio_player.c:460:(.text+0x12a1): undefined reference to `swr_free'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/decoder.c.obj: in function `sc_decoder_open':
E:/scrcpy/scrcpy/app/src/decoder.c:22:(.text+0x6e): undefined reference to `sc_frame_source_sinks_open'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/decoder.c.obj: in function `sc_decoder_close':
E:/scrcpy/scrcpy/app/src/decoder.c:34:(.text+0xbf): undefined reference to `sc_frame_source_sinks_close'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/decoder.c.obj: in function `sc_decoder_push':
E:/scrcpy/scrcpy/app/src/decoder.c:66:(.text+0x1d7): undefined reference to `sc_frame_source_sinks_push'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/decoder.c.obj: in function `sc_decoder_init':
E:/scrcpy/scrcpy/app/src/decoder.c:100:(.text+0x2ba): undefined reference to `sc_frame_source_init'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/delay_buffer.c.obj: in function `run_buffering':
E:/scrcpy/scrcpy/app/src/delay_buffer.c:88:(.text+0x51c): undefined reference to `sc_frame_source_sinks_push'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/delay_buffer.c.obj: in function `sc_delay_buffer_frame_sink_open':
E:/scrcpy/scrcpy/app/src/delay_buffer.c:138:(.text+0x75d): undefined reference to `sc_frame_source_sinks_open'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/delay_buffer.c:151:(.text+0x7b8): undefined reference to `sc_frame_source_sinks_close'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/delay_buffer.c.obj: in function `sc_delay_buffer_frame_sink_close':
E:/scrcpy/scrcpy/app/src/delay_buffer.c:174:(.text+0x887): undefined reference to `sc_frame_source_sinks_close'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/delay_buffer.c.obj: in function `sc_delay_buffer_frame_sink_push':
E:/scrcpy/scrcpy/app/src/delay_buffer.c:199:(.text+0x989): undefined reference to `sc_frame_source_sinks_push'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/delay_buffer.c.obj: in function `sc_delay_buffer_init':
E:/scrcpy/scrcpy/app/src/delay_buffer.c:235:(.text+0xc4e): undefined reference to `sc_frame_source_init'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/demuxer.c.obj: in function `run_demuxer':
E:/scrcpy/scrcpy/app/src/demuxer.c:155:(.text+0x434): undefined reference to `sc_packet_source_sinks_disable'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/demuxer.c:170:(.text+0x4aa): undefined reference to `sc_packet_source_sinks_disable'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/demuxer.c:178:(.text+0x4ef): undefined reference to `sc_packet_source_sinks_disable'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/demuxer.c:217:(.text+0x63d): undefined reference to `sc_packet_source_sinks_open'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/demuxer.c:254:(.text+0x716): undefined reference to `sc_packet_source_sinks_push'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/demuxer.c:270:(.text+0x77f): undefined reference to `sc_packet_source_sinks_close'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/demuxer.c.obj: in function `sc_demuxer_init':
E:/scrcpy/scrcpy/app/src/demuxer.c:287:(.text+0x828): undefined reference to `sc_packet_source_init'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/file_pusher.c.obj: in function `run_file_pusher':
E:/scrcpy/scrcpy/app/src/file_pusher.c:139:(.text+0x9f3): undefined reference to `sc_adb_install'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/file_pusher.c:147:(.text+0xa7d): undefined reference to `sc_adb_push'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/scrcpy.c.obj: in function `scrcpy':
E:/scrcpy/scrcpy/app/src/scrcpy.c:491:(.text+0xe05): undefined reference to `sc_packet_source_add_sink'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/scrcpy.c:496:(.text+0xe4d): undefined reference to `sc_packet_source_add_sink'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/scrcpy.c:517:(.text+0xf3d): undefined reference to `sc_packet_source_add_sink'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/scrcpy.c:521:(.text+0xf72): undefined reference to `sc_packet_source_add_sink'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/scrcpy.c:700:(.text+0x1361): undefined reference to `sc_frame_source_add_sink'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/scrcpy.c:709:(.text+0x13be): undefined reference to `sc_frame_source_add_sink'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/scrcpy.c:715:(.text+0x1425): undefined reference to `sc_frame_source_add_sink'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/server.c.obj: in function `push_server':
E:/scrcpy/scrcpy/app/src/server.c:130:(.text+0x5b4): undefined reference to `sc_adb_push'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/server.c.obj: in function `execute_server':
E:/scrcpy/scrcpy/app/src/server.c:212:(.text+0x847): undefined reference to `sc_adb_get_executable'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/server.c:383:(.text+0x1674): undefined reference to `sc_adb_execute'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/server.c.obj: in function `sc_server_init':
E:/scrcpy/scrcpy/app/src/server.c:483:(.text+0x19fc): undefined reference to `sc_adb_tunnel_init'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/server.c.obj: in function `sc_server_connect_to':
E:/scrcpy/scrcpy/app/src/server.c:609:(.text+0x1e85): undefined reference to `sc_adb_tunnel_close'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/server.c:653:(.text+0x2082): undefined reference to `sc_adb_tunnel_close'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/server.c.obj: in function `get_adb_tcp_port':
E:/scrcpy/scrcpy/app/src/server.c:680:(.text+0x2130): undefined reference to `sc_adb_getprop'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/server.c.obj: in function `sc_server_switch_to_tcpip':
E:/scrcpy/scrcpy/app/src/server.c:746:(.text+0x235a): undefined reference to `sc_adb_get_device_ip'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/server.c:758:(.text+0x23ee): undefined reference to `sc_adb_tcpip'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/server.c.obj: in function `sc_server_connect_to_tcpip':
E:/scrcpy/scrcpy/app/src/server.c:789:(.text+0x24f4): undefined reference to `sc_adb_disconnect'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/server.c:793:(.text+0x2525): undefined reference to `sc_adb_connect'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/server.c.obj: in function `sc_server_configure_tcpip_unknown_address':
E:/scrcpy/scrcpy/app/src/server.c:823:(.text+0x264a): undefined reference to `sc_adb_device_get_type'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/server.c.obj: in function `sc_server_kill_adb_if_requested':
E:/scrcpy/scrcpy/app/src/server.c:849:(.text+0x2768): undefined reference to `sc_adb_kill_server'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/scrcpy.dir/server.c.obj: in function `run_server':
E:/scrcpy/scrcpy/app/src/server.c:862:(.text+0x27a5): undefined reference to `sc_adb_start_server'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/server.c:906:(.text+0x29a6): undefined reference to `sc_adb_select_device'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/server.c:915:(.text+0x2a23): undefined reference to `sc_adb_device_destroy'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/server.c:925:(.text+0x2a91): undefined reference to `sc_adb_device_destroy'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/server.c:966:(.text+0x2cb4): undefined reference to `sc_adb_tunnel_open'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/server.c:976:(.text+0x2d13): undefined reference to `sc_adb_tunnel_close'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/server.c:989:(.text+0x2d96): undefined reference to `sc_adb_tunnel_close'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/msys64/mingw64/lib/libSDL2main.a(SDL_windows_main.c.obj):(.text+0x17a): undefined reference to `SDL_main'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(log.c.obj): in function `sc_set_log_level':
E:/scrcpy/scrcpy/app/src/util/log.c:50:(.text+0x166): undefined reference to `SDL_LogSetPriority'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/util/log.c:51:(.text+0x175): undefined reference to `SDL_LogSetPriority'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(log.c.obj): in function `sc_get_log_level':
E:/scrcpy/scrcpy/app/src/util/log.c:56:(.text+0x18e): undefined reference to `SDL_LogGetPriority'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(log.c.obj): in function `sc_log':
E:/scrcpy/scrcpy/app/src/util/log.c:66:(.text+0x1ea): undefined reference to `SDL_LogMessageV'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(log.c.obj): in function `sc_av_log_callback':
E:/scrcpy/scrcpy/app/src/util/log.c:124:(.text+0x3f7): undefined reference to `SDL_LogMessageV'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(log.c.obj): in function `sc_log_configure':
E:/scrcpy/scrcpy/app/src/util/log.c:151:(.text+0x4d2): undefined reference to `SDL_LogSetOutputFunction'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/util/log.c:153:(.text+0x4e1): undefined reference to `av_log_set_callback'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(net.c.obj): in function `net_init':
E:/scrcpy/scrcpy/app/src/util/net.c:33:(.text+0x1f): undefined reference to `__imp_WSAStartup'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(net.c.obj): in function `net_cleanup':
E:/scrcpy/scrcpy/app/src/util/net.c:45:(.text+0x71): undefined reference to `__imp_WSACleanup'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(net.c.obj): in function `wrap':
E:/scrcpy/scrcpy/app/src/util/net.c:59:(.text+0xe7): undefined reference to `__imp_closesocket'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(net.c.obj): in function `sc_raw_socket_close':
E:/scrcpy/scrcpy/app/src/util/net.c:91:(.text+0x163): undefined reference to `__imp_closesocket'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(net.c.obj): in function `net_perror':
E:/scrcpy/scrcpy/app/src/util/net.c:119:(.text+0x1d4): undefined reference to `__imp_WSAGetLastError'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(net.c.obj): in function `net_socket':
E:/scrcpy/scrcpy/app/src/util/net.c:130:(.text+0x20a): undefined reference to `__imp_socket'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(net.c.obj): in function `net_connect':
E:/scrcpy/scrcpy/app/src/util/net.c:150:(.text+0x2a5): undefined reference to `__imp_htonl'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/util/net.c:151:(.text+0x2b7): undefined reference to `__imp_htons'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/util/net.c:153:(.text+0x2d5): undefined reference to `__imp_connect'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(net.c.obj): in function `net_listen':
E:/scrcpy/scrcpy/app/src/util/net.c:166:(.text+0x356): undefined reference to `__imp_setsockopt'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/util/net.c:173:(.text+0x37e): undefined reference to `__imp_htonl'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/util/net.c:174:(.text+0x390): undefined reference to `__imp_htons'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/util/net.c:176:(.text+0x3ae): undefined reference to `__imp_bind'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/util/net.c:181:(.text+0x3dc): undefined reference to `__imp_listen'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(net.c.obj): in function `net_accept':
E:/scrcpy/scrcpy/app/src/util/net.c:201:(.text+0x440): undefined reference to `__imp_accept'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(net.c.obj): in function `net_recv':
E:/scrcpy/scrcpy/app/src/util/net.c:214:(.text+0x4ca): undefined reference to `__imp_recv'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(net.c.obj): in function `net_recv_all':
E:/scrcpy/scrcpy/app/src/util/net.c:220:(.text+0x519): undefined reference to `__imp_recv'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(net.c.obj): in function `net_send':
E:/scrcpy/scrcpy/app/src/util/net.c:226:(.text+0x568): undefined reference to `__imp_send'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(net.c.obj): in function `net_interrupt':
E:/scrcpy/scrcpy/app/src/util/net.c:252:(.text+0x65b): undefined reference to `__imp_closesocket'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(net.c.obj): in function `net_close':
E:/scrcpy/scrcpy/app/src/util/net.c:267:(.text+0x6b3): undefined reference to `__imp_closesocket'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(net.c.obj): in function `net_parse_ipv4':
E:/scrcpy/scrcpy/app/src/util/net.c:279:(.text+0x6fd): undefined reference to `__imp_inet_pton'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:/scrcpy/scrcpy/app/src/util/net.c:284:(.text+0x731): undefined reference to `__imp_ntohl'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(thread.c.obj): in function `sc_thread_create':
E:/scrcpy/scrcpy/app/src/util/thread.c:16:(.text+0x70): undefined reference to `SDL_CreateThread'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(thread.c.obj): in function `sc_thread_set_priority':
E:/scrcpy/scrcpy/app/src/util/thread.c:50:(.text+0x15d): undefined reference to `SDL_SetThreadPriority'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(thread.c.obj): in function `sc_thread_join':
E:/scrcpy/scrcpy/app/src/util/thread.c:61:(.text+0x1b7): undefined reference to `SDL_WaitThread'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(thread.c.obj): in function `sc_mutex_init':
E:/scrcpy/scrcpy/app/src/util/thread.c:66:(.text+0x1cf): undefined reference to `SDL_CreateMutex'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(thread.c.obj): in function `sc_mutex_destroy':
E:/scrcpy/scrcpy/app/src/util/thread.c:81:(.text+0x25b): undefined reference to `SDL_DestroyMutex'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(thread.c.obj): in function `sc_mutex_lock':
E:/scrcpy/scrcpy/app/src/util/thread.c:88:(.text+0x2b3): undefined reference to `SDL_LockMutex'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(thread.c.obj): in function `sc_mutex_unlock':
E:/scrcpy/scrcpy/app/src/util/thread.c:108:(.text+0x36b): undefined reference to `SDL_UnlockMutex'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(thread.c.obj): in function `sc_thread_get_id':
E:/scrcpy/scrcpy/app/src/util/thread.c:121:(.text+0x3a9): undefined reference to `SDL_ThreadID'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(thread.c.obj): in function `sc_cond_init':
E:/scrcpy/scrcpy/app/src/util/thread.c:135:(.text+0x3f8): undefined reference to `SDL_CreateCond'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(thread.c.obj): in function `sc_cond_destroy':
E:/scrcpy/scrcpy/app/src/util/thread.c:147:(.text+0x468): undefined reference to `SDL_DestroyCond'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(thread.c.obj): in function `sc_cond_wait':
E:/scrcpy/scrcpy/app/src/util/thread.c:152:(.text+0x495): undefined reference to `SDL_CondWait'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(thread.c.obj): in function `sc_cond_timedwait':
E:/scrcpy/scrcpy/app/src/util/thread.c:176:(.text+0x563): undefined reference to `SDL_CondWaitTimeout'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(thread.c.obj): in function `sc_cond_signal':
E:/scrcpy/scrcpy/app/src/util/thread.c:194:(.text+0x635): undefined reference to `SDL_CondSignal'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: util/libutil.a(thread.c.obj): in function `sc_cond_broadcast':
E:/scrcpy/scrcpy/app/src/util/thread.c:207:(.text+0x681): undefined reference to `SDL_CondBroadcast'
collect2.exe: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.


what can i do？

**Comment by rom1v** [comment]
Sorry, scrcpy is intended to be built with meson+ninja. If you decide to hack and write your own scripts to use another build system, I can't help.
